### PR TITLE
[4.1] Workflow Plugin Image Settings

### PR DESCRIPTION
**Pull request by david0296**
Pull Request for Issue # .

### Summary of Changes
Added new options on the transition settings to set an image of an article (Intro Image + Full Article Image) as required, set a resolution for resizing and set allowed file extensions.

image

### Testing Instructions
1. Navigate to Content -> Workflow -> Transitions
2. Edit a transition of your choise
3. Go to the tab "Image Settings", choose your settings and save
4. Navigate to Content -> Articles and execute the edited transition
5. Make sure, the transition is not executed, if one of the images, which is set to be required, is missing
6. Make sure, the transition is not executed, if one of the images has a not-supported file extension
7. Make sure, there is a resized image in the image folder with the correct resolution

### CODE REVIEW

File: administrator/language/en-GB/plg_workflow_images.ini [hunk 1]
```diff
@@ -0,0 +1,22 @@
+; Joomla! Project
+; Copyright (C) 2005 - 2020 Open Source Matters. All rights reserved.
+; License GNU General Public License version 2 or later; see LICENSE.txt, see LICENSE.php
+; Note : All ini files need to be saved as UTF-8
```
Comment (chmst) [inline administrator/language/en-GB/plg_workflow_images.ini:4]: Please alphaorder language keys

**Comment by brianteeman** [comment]
For Joomla 4 we have tried to remove any description text that is not really necessary. Please could you review your strings with that in mind.

File: administrator/language/en-GB/plg_workflow_images.ini (hunk 1 above)
Comment (david0296) [reply administrator/language/en-GB/plg_workflow_images.ini:4]: done

**Comment by david0296** [comment]
@brianteeman thanks for this information. I will take another look at it

**Comment by david0296** [comment]
I've added two radio buttons on the form, so you can turn on/of, if the Intro Image/Full Article Image gets resized or not after executing the transition

**Comment by SharkyKZ** [comment]
Is this intentionally limited to local filesystem?

File: plugins/workflow/images/images.php [hunk 2]
```diff
@@ -0,0 +1,442 @@
...
+				if (!($fullArticleImagePath || !file_exists($fullArticleImagePath)))
+				{
+					Factory::getApplication()->enqueueMessage(Text::_('PLG_WORKFLOW_IMAGES_FULL_ARTICLE_IMAGE_REQUIRED'));
+					$event->setStopTransition();
+
+					return false;
+				}
+			}
+
+			if( !exif_imagetype($introImagePath)){
```
Comment (SharkyKZ) [inline plugins/workflow/images/images.php]: We don't have a dependency on exif. There's already `Joomla\CMS\Helper\MediaHelper::getMimeType()` for checking mime type.

Comment (david0296) [reply plugins/workflow/images/images.php]: Thank you, i have changed it

**Comment by david0296** [comment]
> Is this intentionally limited to local filesystem?

I'm sorry, but I don't know how to test this/how to set an external image as "Intro Image" or "Full Article Image"

File: administrator/language/en-GB/plg_workflow_images.ini [hunk 3]
```diff
@@ -0,0 +1,21 @@
+; Joomla! Project
+; Copyright (C) 2005 - 2020 Open Source Matters. All rights reserved.
```
Comment (PhilETaylor) [inline administrator/language/en-GB/plg_workflow_images.ini:2]:
```suggestion
; (C) 2021 Open Source Matters, Inc. <https://www.joomla.org>
```

File: administrator/language/en-GB/plg_workflow_images.sys.ini [hunk 4]
```diff
@@ -0,0 +1,7 @@
+; Joomla! Project
+; Copyright (C) 2005 - 2020 Open Source Matters. All rights reserved.
```
Comment (PhilETaylor) [inline administrator/language/en-GB/plg_workflow_images.sys.ini:2]:
```suggestion
; (C) 2021 Open Source Matters, Inc. <https://www.joomla.org>
```

File: administrator/language/en-GB/plg_workflow_images.ini [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+; License GNU General Public License version 2 or later; see LICENSE.txt, see LICENSE.php
```
Comment (PhilETaylor) [inline administrator/language/en-GB/plg_workflow_images.ini:3]:
```suggestion
; License GNU General Public License version 2 or later; see LICENSE.txt
```

File: administrator/language/en-GB/plg_workflow_images.sys.ini [hunk 6]
```diff
@@ -0,0 +1,7 @@
...
+; License GNU General Public License version 2 or later; see LICENSE.txt, see LICENSE.php
```
Comment (PhilETaylor) [inline administrator/language/en-GB/plg_workflow_images.sys.ini:3]:
```suggestion
; License GNU General Public License version 2 or later; see LICENSE.txt
```

File: plugins/workflow/images/images.php [hunk 7]
```diff
@@ -0,0 +1,430 @@
+<?php
+/**
+ * @package     Joomla.Plugin
+ * @subpackage  Workflow.Publishing
+ *
+ * @copyright   Copyright (C) 2005 - 2020 Open Source Matters, Inc. All rights reserved.
```
Comment (PhilETaylor) [inline plugins/workflow/images/images.php:7]:
```suggestion
 * @copyright   (C) 2021 Open Source Matters, Inc. <https://www.joomla.org>
```

File: plugins/workflow/images/images.php [hunk 8]
```diff
@@ -0,0 +1,430 @@
...
+ */
+class PlgWorkflowImages extends CMSPlugin implements SubscriberInterface
+{
+	use WorkflowPluginTrait;
+
+	/**
+	 * Load the language file on instantiation.
+	 *
+	 * @var    boolean
+	 * @since  4.0.0
```
Comment (PhilETaylor) [inline plugins/workflow/images/images.php]:
```suggestion
	 * @since  __DEPLOY_VERSION__
```

File: plugins/workflow/images/images.php [hunk 9]
```diff
@@ -0,0 +1,430 @@
...
+	 */
+	protected $autoloadLanguage = true;
+
+	/**
+	 * Loads the CMS Application for direct access
+	 *
+	 * @var   CMSApplicationInterface
+	 * @since 4.0.0
```
Comment (PhilETaylor) [inline plugins/workflow/images/images.php]:
```suggestion
	 * @since __DEPLOY_VERSION__
```

File: plugins/workflow/images/images.php [hunk 10]
```diff
@@ -0,0 +1,430 @@
...
+	 */
+	protected $app;
+
+	/**
+	 * The name of the supported name to check against
+	 *
+	 * @var   string
+	 * @since 4.0.0
```
Comment (PhilETaylor) [inline plugins/workflow/images/images.php]:
```suggestion
	 * @since __DEPLOY_VERSION__
```

File: plugins/workflow/images/images.php [hunk 11]
```diff
@@ -0,0 +1,430 @@
...
+	 */
+	protected $supportFunctionality = 'core.state';
+
+	/**
+	 * Returns an array of events this subscriber will listen to.
+	 *
+	 * @return  array
+	 *
+	 * @since   4.0.0
```
Comment (PhilETaylor) [inline plugins/workflow/images/images.php]:
```suggestion
	 * @since   __DEPLOY_VERSION__
```

File: plugins/workflow/images/images.php [hunk 12]
```diff
@@ -0,0 +1,430 @@
...
+			'onWorkflowFunctionalityUsed' => 'onWorkflowFunctionalityUsed',
+		];
+	}
+
+	/**
+	 * The form event.
+	 *
+	 * @param   EventInterface  $event  The event
+	 *
+	 * @since   4.0.0
```
Comment (PhilETaylor) [inline plugins/workflow/images/images.php]:
```suggestion
	 * @since   __DEPLOY_VERSION__
```

File: plugins/workflow/images/images.php [hunk 13]
```diff
@@ -0,0 +1,430 @@
...
+	 *
+	 * @return boolean
+	 *
+	 * @throws Exception
+	 * @throws Exception
+	 * @throws Exception
+	 * @throws Exception
+	 * @throws Exception
+	 * @throws Exception
+	 * @since   4.0.0
```
Comment (PhilETaylor) [inline plugins/workflow/images/images.php]:
```suggestion
	 * 
	 * @since   __DEPLOY_VERSION__
```

File: plugins/workflow/images/images.php [hunk 14]
```diff
@@ -0,0 +1,430 @@
...
+	 * @param WorkflowTransitionEvent $event
+	 *
+	 * @return boolean
+	 *
+	 * @throws Exception
+	 * @throws Exception
+	 * @throws Exception
+	 * @throws Exception
+	 * @throws Exception
+	 * @throws Exception
```
Comment (PhilETaylor) [inline plugins/workflow/images/images.php]:
```suggestion	
	 * @throws Exception
```

File: plugins/workflow/images/images.php [hunk 15]
```diff
@@ -0,0 +1,430 @@
...
+	 * @since   4.0.0
+	 */
+	public function onWorkflowBeforeTransition(WorkflowTransitionEvent $event)
+	{
+		$context    = $event->getArgument('extension');
+		$extensionName = $event->getArgument('extensionName');
+		$transition = $event->getArgument('transition');
+		$pks        = $event->getArgument('pks');
+
+		//get Values from Form
```
Comment (PhilETaylor) [inline plugins/workflow/images/images.php]:
```suggestion
		// Get the values from Form
```

File: plugins/workflow/images/images.php [hunk 16]
```diff
@@ -0,0 +1,430 @@
...
+		$introImageRequired = $transition->options->get('images_intro_image_settings');
+		$fullArticleImageRequired = $transition->options->get('images_full_article_image_settings');
+		//$allowedExtensions = str_replace(',','|',$transition->options->get('fileExtensions'));
```
Comment (PhilETaylor) [inline plugins/workflow/images/images.php]:
```suggestion
```

File: plugins/workflow/images/images.php [hunk 17]
```diff
@@ -0,0 +1,430 @@
...
+
```
Comment (PhilETaylor) [inline plugins/workflow/images/images.php]:
```suggestion
```

File: plugins/workflow/images/images.php [hunk 18]
```diff
@@ -0,0 +1,430 @@
...
+		$component = $this->app->bootComponent($extensionName);
+
+		$options = [
+			'ignore_request'            => true,
+			// We already have triggered onContentBeforeChangeState, so use our own
+			'event_before_change_state' => 'onWorkflowBeforeChangeState'
+		];
+
+		// Get model
+
```
Comment (PhilETaylor) [inline plugins/workflow/images/images.php]:
```suggestion
```

File: plugins/workflow/images/images.php [hunk 19]
```diff
@@ -0,0 +1,430 @@
...
+		$modelName = $component->getModelName($context);
+
+		$model = $component->getMVCFactory()->createModel($modelName, $this->app->getName(), $options);
+
+
+
```
Comment (PhilETaylor) [inline plugins/workflow/images/images.php]:
```suggestion
```

File: plugins/workflow/images/images.php [hunk 20]
```diff
@@ -0,0 +1,430 @@
...
+	}
+
+	/**
+	 * Change State of an item. Used to disable state change
+	 *
+	 * @param   WorkflowTransitionEvent  $event
+	 *
+	 * @return boolean
+	 *
+	 * @since   4.0.0
```
Comment (PhilETaylor) [inline plugins/workflow/images/images.php]:
```suggestion
	 * @since   __DEPLOY_VERSION__
```

File: plugins/workflow/images/images.php [hunk 21]
```diff
@@ -0,0 +1,430 @@
...
+
+	/**
+	 * Change State of an item. Used to disable state change
+	 *
+	 * @param   EventInterface  $event
+	 *
+	 * @return boolean
+	 *
+	 * @throws Exception
+	 * @since   4.0.0
```
Comment (PhilETaylor) [inline plugins/workflow/images/images.php]:
```suggestion
	 * @since   __DEPLOY_VERSION__
```

File: plugins/workflow/images/images.php [hunk 22]
```diff
@@ -0,0 +1,430 @@
...
+	}
+
+	/**
+	 * The save event.
+	 *
+	 * @param   EventInterface  $event
+	 *
+	 * @return  boolean
+	 *
+	 * @since   4.0.0
```
Comment (PhilETaylor) [inline plugins/workflow/images/images.php]:
```suggestion
	 * @since   __DEPLOY_VERSION__
```

File: plugins/workflow/images/images.php [hunk 23]
```diff
@@ -0,0 +1,430 @@
...
+	}
+
+	/**
+	 * Check if the current plugin should execute workflow related activities
+	 *
+	 * @param   string  $context
+	 *
+	 * @return boolean
+	 *
+	 * @since   4.0.0
```
Comment (PhilETaylor) [inline plugins/workflow/images/images.php]:
```suggestion
	 * @since   __DEPLOY_VERSION__
```

File: plugins/workflow/images/images.php [hunk 24]
```diff
@@ -0,0 +1,430 @@
...
+
+		return true;
+	}
+
+	/**
+	 * If plugin supports the functionality we set the used variable
+	 *
+	 * @param   WorkflowFunctionalityUsedEvent  $event
+	 *
+	 * @since 4.0.0
```
Comment (PhilETaylor) [inline plugins/workflow/images/images.php]:
```suggestion
	 * @since __DEPLOY_VERSION__
```

File: plugins/workflow/images/images.xml [hunk 25]
```diff
@@ -0,0 +1,43 @@
+<?xml version="1.0" encoding="utf-8"?>
+<extension type="plugin" group="workflow" method="upgrade">
+	<name>plg_workflow_images</name>
+	<author>Joomla! Project</author>
+	<creationDate>March 2020</creationDate>
+	<copyright>Copyright (C) 2005 - 2020 Open Source Matters. All rights reserved.</copyright>
```
Comment (PhilETaylor) [inline plugins/workflow/images/images.xml:6]:
```suggestion
	<copyright>Copyright (C) 2021 Open Source Matters, Inc. <https://www.joomla.org></copyright>
```

File: plugins/workflow/images/images.xml [hunk 26]
```diff
@@ -0,0 +1,43 @@
+<?xml version="1.0" encoding="utf-8"?>
+<extension type="plugin" group="workflow" method="upgrade">
+	<name>plg_workflow_images</name>
+	<author>Joomla! Project</author>
+	<creationDate>March 2020</creationDate>
```
Comment (PhilETaylor) [inline plugins/workflow/images/images.xml:5]: needs updating eventually :)

File: plugins/workflow/images/images.xml [hunk 27]
```diff
@@ -0,0 +1,43 @@
...
+	<copyright>Copyright (C) 2005 - 2020 Open Source Matters. All rights reserved.</copyright>
+	<license>GNU General Public License version 2 or later; see LICENSE.txt</license>
+	<authorEmail>admin@joomla.org</authorEmail>
+	<authorUrl>www.joomla.org</authorUrl>
+	<version>4.0.0</version>
```
Comment (PhilETaylor) [inline plugins/workflow/images/images.xml:10]:
```suggestion
	<version>__DEPLOY_VERSION__</version>
```

**Comment by HLeithner** [comment]
This pull request has automatically rebased to 4.2-dev.

**Comment by rdeutz** [comment]
Seems to be a very special usecase and not so important because nobody worked on it for a longer time. I am closing it. Thanks for the idea and working on the implementation.

**Pull request closed without merging**